src/eval.rs

```diff
@@ -6,46 +6,86 @@ use crate::EvalError;
 use gc::{Finalize, Gc, GcCell, Trace};
 use rnix::TextRange;
 use std::borrow::Borrow;
+use std::collections::HashMap;
 
-type ExprResult = Result<Box<Expr>, EvalError>;
+// Expressions like BinOp have the only copy of their Expr children,
+// so they use ExprResultBox. Expressions like Map, which may have
+// contents copied in multiple places, need ExprResultGc.
+type ExprResultBox = Result<Box<Expr>, EvalError>;
+type ExprResultGc = Result<Gc<Expr>, EvalError>;
 
 /// Used to lazily calculate the value of a Expr. This should be
 /// tolerant of parsing and evaluation errors from child Exprs.
-#[derive(Debug, Clone, Trace, Finalize)]
+///
+/// We store everything that we want the user to inspect. For example,
+/// the source for an attribute key-value pair includes the key so the
+/// user can hover inside dynamic keys in code like
+/// `{ "${toString (1+1)}" = 2; }`.
+#[derive(Debug, Trace, Finalize)]
 pub enum ExprSource {
+    // We want to share child MapAttrs between the ExprSource
+    // and the value map, so we use Gc.
+    Map {
+        inherits: Vec<ExprResultGc>,
+        definitions: Vec<ExprResultGc>,
+    },
+    // When we have a key-value pair like `foo = 123`, we put the
+    // key-value pair as the value in the map, then we evaluate that
+    // to just the value. This sounds odd, but it helps the LSP
```

Comment:
I'm not really sure if I understand what you're talking about here. We still keep the keys somehow, correct?

Reply:
Yep, you're right. I spent a lot of time on that comment because it was "hard to explain," but I think I was just wrong :)

I'll try to refactor `MapAttr` to make it cleaner